### ISSUE TITLE
Pytorch Official Models Perform bad in ImageNet Caffe Dataset

### ISSUE THREAD

**Issue by csyhhu** [jiecaoyu/XNOR-Net-PyTorch#62]
Hi,

I download you preprocessing ImageNet lmdb dataset and try to validate using pytorch official model, such as ResNet18 from https://download.pytorch.org/models/resnet18-5c106cde.pth.

I have tried your code using imagenet_mean.binaryproto or pytorch code using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225] as dataset preprocess, but both showed very bad performance, nearly 0.125% for Prec@1. 

Have you tried to use pytorch official model with your preprocssing lmdb dataset? Or do you have any idea how this problem comes out? (For example, how you preprocess the ImageNet dataset into lmdb format.)

Looking forward to your reply. Thanks!

Best regards.

**Comment by jiecaoyu** [comment]
@csyhhu Hi, the preprocessing of Caffe is different from PyTorch. I think you might get better accuracy number if you multiply both ```mean``` and ```std``` with 255. Current scripts support training in PyTorch.

**Comment by jiecaoyu** [comment]
BTW, the training using torchvision has slightly worse accuracy than using the Caffe dataset. I don't have much computation resources to finetune the hyperparameters.

**Comment by csyhhu** [comment]
@jiecaoyu Hi! Thanks for your prompt reply! My current case is that I want to use the lmdb dataset to accelerate training, with a pretrained model. Now I can download pytorch official model but it seems that it doesn't work well on your preprocessed lmdb dataset. 

There should be some difference between lmdb dataset and pytorch one (who uses raw images). For example lmdb only contains a mean file while pytorch has std.

I have tried to multiply 255 to pytorch's mean and st, and the performance does climb a lot, but still show some differences. It seems that caffe's input is ranged in [-255, 255] while pytorch's inputs is [-1, 1]. Thanks for your help!

One more question, did you use the default conver_imagenetset.cpp in caffe to generate the lmdb dataset with RESIZE=TRUE?

**Comment by jiecaoyu** [comment]
@csyhhu Yes, all images are resized to 256x256

**Comment by csyhhu** [comment]
Hi @jiecaoyu ,

I may find out a way to use Caffe-generated lmdb (your uploaded one) data on pytorch official models. As is mentioned in https://pytorch.org/docs/stable/torchvision/models.html, pytorch's images should be ranged at [0,1], then use normalization. However, data in lmdb is ranged at [0, 255]. Therefore I divide by 255. when data is extracted from lmdb. 

However, it seems that your generated data is formatted in BGR, while pytorch's images are RGB -- in fact it depends on how to read image, pytorch uses PIL for example. Therefore a channel permutation is needed in data reading.

After doing all this, performance climbs up to 54.146/77.464, still shows 1~2% gap between the released report of pytorch. But I think it's quite close, after all, using lmdb data can much increase data loading, at least for my server.

Besides, I find that there is more efficient way to read lmdb dataset without using caffe. NVIDIA provides an interface to do so at: https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/dataloading_lmdb.html, you may try this if needed.

If you want a complete change of code, please leave an email address I can send it to you.

Thanks again for providing the code !

**Comment by jiecaoyu** [comment]
@csyhhu Thanks for your reply! If your CPU is not too old you can checkout this [link](https://github.com/pytorch/vision#image-backend). By using either [Pillow-SIMD](https://github.com/uploadcare/pillow-simd) or [accimage](https://github.com/pytorch/accimage)， you can use SIMD support to accelerate the torchvision support. It accelerates the preprocessing a lot. I do **not** recommend Pillow-SIMD since it seems has some installation problem. You can try accimage first.
